### ISSUE TITLE
First build succeeds but later ones fail if HOME=/project/path

### ISSUE THREAD

**Issue by sp3d** [rust-lang/cargo#1731]
I tend to set $HOME to the path of current projects to ensure that state isn't accidentally shared between multiple projects. This seems to lead Cargo to erroneously and misleadingly complain, "Could not find `Cargo.toml` in `.cargo/registry/src/github.com-0a35038f75765ae4/rand-0.3.8/rand_macros`" when compiling projects that use the `rand` crate (or, presumably, any crate with similar Cargo setup).

This script demonstrates the problem:
```
#!/bin/sh
# reproduces "Could not find `Cargo.toml` in `.cargo/registry/src/github.com-0a35038f75765ae4/rand-0.3.8/rand_macros`" bug
set -ex

mkdir reproducing-cargo-bug
cd reproducing-cargo-bug

echo '[package]
name = "bugrepro"
version = "0.1.0"
authors = [""]

[dependencies]
rand = "*"' > Cargo.toml
mkdir src
touch src/lib.rs
echo "building once!"
HOME=`pwd` cargo build --verbose #works
echo "building again!"
HOME=`pwd` cargo build --verbose #fails!
```

**Comment by alexcrichton** [comment]
I'm unable to reproduce this any of nightly/beta/stable, perhaps there's something else awry locally? What version of Cargo are you using?

**Comment by sp3d** [comment]
```
$ cargo --version
cargo 0.3.0-nightly (14e8ed9 2015-06-15) (built 2015-06-16)
```
But I've also seen this with previous Cargo nightlies; it occurs on 3 different machines of the same configuration: Arch Linux, x86-64, with Cargo and Rust from the "rust-nightly-bin" package in the AUR.

**Comment by alexcrichton** [comment]
Ah ok I see what's going on here, the problem is that the repos downloaded from crates.io are not intended to be valid standalone project directories (e.g. path dependencies are not uploaded, they're translated to crates.io dependencies). The problem here is that the downloads are all stored in the current directory, and Cargo by default crawls the current directory looking for packages, so it'll recursively go into this directory and then attempt to load an invalid package.

Long story short this is a consequence of https://github.com/rust-lang/cargo/issues/1524, so I'm going to close this in favor of that. Thanks for the report regardless though!
